### PR TITLE
Support manifest files for pacman

### PR DESCRIPTION
**Pull request by Foxboron**
This implements manifest files for pacman with some awk-ward code.

~~Also implements support for `shell` in `run` as the code would always split up the string.~~

### CODE REVIEW

**Review by behrmann — COMMENTED** [review]
Nice! With this all major distros support the manifest. :)

File: mkosi/backend.py [hunk 1]
```diff
@@ -790,13 +790,16 @@ def spawn(
...
     delay_interrupt: bool = True,
     stdout: _FILE = None,
     stderr: _FILE = None,
     env: Mapping[str, Any] = {},
     **kwargs: Any,
 ) -> CompletedProcess:
-    cmdline = [str(x) for x in cmdline]
+
+    if not shell:
+        cmdline = [str(x) for x in cmdline]
```
[the remark this answers is not in the document]
Comment (Foxboron) [reply mkosi/backend.py:802]: Yes, and with this we should not hit the case where we pass a list with `shell=True`

[the remark this answers is not in the document]
Comment (behrmann) [reply mkosi/backend.py:802]: Ah, sorry, I was misunderstanding this, since I'm mostly only used to passing lists to `subproces.run` and that's what usually (always?) happening in the rest of the codebase.

How about?
```suggestion
    cmdline = [str(x) for x in cmdline]
    if shell and isinstance(cmdline, list):
        cmdline = " ".join(t for t in cmdline))
```

Normally I'd throw in a `shlex.quote` for good measure, too, but this would preclude the use with globs that you want.

[the remark this answers is not in the document]
Comment (Foxboron) [reply mkosi/backend.py:802]: `isinstance(cmdline, list)` would always be true. I don't see much use in joining the list again when we get the string itself passed as a parameter.

File: mkosi/manifest.py [hunk 2]
```diff
@@ -197,6 +199,27 @@ def record_deb_packages(self, root: Path) -> None:
 
             source_package.add(package)
 
+    def record_pkg_packages(self, root: Path) -> None:
+        c = run(f'awk \'/%NAME%/ {{getline; printf("%s ", $1)}}; /%VERSION%/ {{getline;printf("%s ", $1)}}; /%BASE%/ {{getline;printf("%s ", $1)}}; /%ARCH%/ {{getline; printf("%s\\n", $1)}}\' {root}/var/lib/pacman/local/*/desc',
+            stdout=PIPE,
+            text=True,
+            shell=True
+        )
```
[the remark this answers is not in the document]
Comment (Foxboron) [reply mkosi/manifest.py:207]: Yep, this makes sense. The `awk` stuff isn't great and pacman is hopefully getting support for `-Q --print-format` in the future.

File: mkosi/backend.py (hunk 1 above)
[the remark this answers is not in the document]
Comment (behrmann) [reply mkosi/backend.py:802]: Sorry, I didn't properly explain my thinking. 

I quickly grepped through the codebase and didn't find  an instance of `run` not being called with a list as an argument. I might have missed something, but the vast majority of call sites are like this. This can't be really encoded on the type level, though, since the `cmdline` argument for `run` is `Sequence[str]` (it's actually not, but I want to simplify this), which was chosen because sequences are immutable by default and it's nice to be general about what one accepts (also it's hard to just change this to list, because lists are invariante and sequenes are covariant so mypy complains about this). 

Unfortunately this is a bit of a footgun, because `str` is also a `Sequence[str]`, so this doesn't ensure the assumption of something that's a bag of words coming in.

Long story short: I was always thinking about this with a list going in, and in that case the join is needed to make `shell=True` work. 

You're right about the `isinstance` there being pointless, I tacked that on without thinking, because I wanted to make the case of a string being passed in work, but in any case keeping a list as input for `run` makes the most sense overall.

**Comment by Foxboron** [comment]
I went with the code suggested by @behrmann and dropped the `shell` argument for the sake of keeping it simple :p

Do you want co-author credit on the commit @behrmann?

File: mkosi/manifest.py [hunk 3]
```diff
@@ -197,6 +199,31 @@ def record_deb_packages(self, root: Path) -> None:
 
             source_package.add(package)
 
+    def record_pkg_packages(self, root: Path) -> None:
+        package_descriptions = [str(p) for p in Path("/var/lib/pacman/local").glob("*/desc")]
```
Comment (DaanDeMeyer) [inline mkosi/manifest.py:203]: Shouldn't we be looking inside `root` here? This will operate on the host system no?

Comment (Foxboron) [reply mkosi/manifest.py:203]: Argh, I didn't tripple check the path. You are correct!

Comment (behrmann) [reply mkosi/manifest.py]: Sorry, my bad, copied that from my local test.

**Comment by behrmann** [comment]
> Do you want co-author credit on the commit @behrmann?

I don't need the laurels, but it might be nice to future people for blameability. You decide.

File: mkosi/manifest.py [hunk 4]
```diff
@@ -197,6 +199,31 @@ def record_deb_packages(self, root: Path) -> None:
 
             source_package.add(package)
 
+    def record_pkg_packages(self, root: Path) -> None:
+        package_descriptions = [str(p) for p in Path(f"{root}/var/lib/pacman/local").glob("*/desc")]
```
Comment (behrmann) [inline mkosi/manifest.py:203]:
```suggestion
        package_descriptions = [str(p) for p in (root / "var/lib/pacman/local").glob("*/desc")]
```
or
```suggestion
        package_descriptions = [str(p) for p in root.joinpath("var/lib/pacman/local").glob("*/desc")]
```
since root already comes in as `Path`.

Comment (Foxboron) [reply mkosi/manifest.py:203]: The slashes are nicer

**Comment by Foxboron** [comment]
>I don't need the laurels, but it might be nice to future people for blameability. You decide.

I'm lazy and would need to google how to do this again :p So can drop it then.

File: mkosi/manifest.py [hunk 5]
```diff
@@ -197,6 +199,31 @@ def record_deb_packages(self, root: Path) -> None:
...
+        package_descriptions = [str(p) for p in (root / "/var/lib/pacman/local").glob("*/desc")]
+        awkprog = """\
+BEGIN { RS="\\n\\n"; FS="\\n"; }
+/%NAME%/ { name=$2 };
+/%VERSION%/ { version=$2 };
+/%BASE%/ { base=$2 };
+/%ARCH%/ { arch=$2 };
+ENDFILE { print name, version, base, arch }
+"""
+        c = run(["awk", awkprog] + package_descriptions, stdout=PIPE, text=True)
```
Comment (poettering) [inline mkosi/manifest.py:212]: hmpf, you shell out to awk for a simple parser like that? Let's maybe not do that... The awk expression above sounds simple enough to just implement in python.

Comment (Foxboron) [reply mkosi/manifest.py]: Fixed :)

**Comment by Foxboron** [comment]
Last iteration removes the awk-ward code :)

File: mkosi/manifest.py [hunk 6]
```diff
@@ -197,6 +214,21 @@ def record_deb_packages(self, root: Path) -> None:
 
             source_package.add(package)
 
+    def record_pkg_packages(self, root: Path) -> None:
+        package_descriptions = [str(p) for p in (root / "var/lib/pacman/local").glob("*/desc")]
+        packages = sorted(package_descriptions)
```
Comment (behrmann) [inline mkosi/manifest.py:219]:
```suggestion
        packages = sorted(root.joinpath("var/lib/pacman/local").glob("*/desc"))
```
should get rid of the mypy error (`package` is a str instead of the expected `Path`)

Comment (Foxboron) [reply mkosi/manifest.py:219]: Clearly I do need to install mypy so I don't need to rely on the CI :/

File: mkosi/manifest.py [hunk 7]
```diff
@@ -64,6 +64,21 @@ def report(self) -> str:
...
+            if line == "%NAME%":
+                ret.append(lines[n+1])
+            if line == "%VERSION%":
+                ret.append(lines[n+1])
+            if line == "%BASE%":
+                ret.append(lines[n+1])
+            if line == "%ARCH%":
+                ret.append(lines[n+1])
+                break
+    return ret
```
Comment (keszybz) [inline mkosi/manifest.py:81]:
```
def parse_pkg_desc(f: Path) -> (str,):
     with f.open() as desc:
        for line in desc:
            match line.strip():
               case '%NAME%':
                  name = next(desc).strip()
               case '%VERSION%':
                  version = next(desc).strip()
               case '%BASE%':
                  base = next(desc).strip()
               case '%ARCH%':
                  arch = next(desc).strip()
                  return name, version, base, arch
```

Untested, but I think that's better.

Comment (keszybz) [reply mkosi/manifest.py:81]: I don't know how reliable those files are. This will "crash" if some field is not found.

Comment (DaanDeMeyer) [reply mkosi/manifest.py:81]: Using `match` isn't an option until we bump the minimum python version unfortunately

Comment (Foxboron) [reply mkosi/manifest.py:81]: I checked if I could use match actually :) but currently `mkosi` is only up to python3.7

Comment (Foxboron) [reply mkosi/manifest.py:81]: @keszybz The files should *always* have these fields present, and in *exactly* this order. There is no spec/manpage, and we can make it more flexible. But this is very well handled in `makepkg`

Comment (behrmann) [reply mkosi/manifest.py:81]: We'd need to bump the Python version quite a bit. ;) We're on 3.7+ and `match` has only been available for less than a year with 3.10. I'm looking forward to using it but maybe we should do a more conservative bump before that. 

But I also prefer the pre 3.10 version of the code (among other reasons because the returned type has a more concrete shape).
```python
def parse_pkg_desc(f: Path) -> Tuple[str, str, str, str]:
    with open(f) as desc:
        for line in desc:
            line = line.strip()
            if line == "%NAME%":
                name = next(desc).strip()
            elif line == "%VERSION%":
                version = next(desc).strip()
            elif line == "%BASE%":
                base = next(desc).strip()
            elif line == "%ARCH%":
                arch = next(desc).strip()
                break
    return name, version, base, break
```

File: mkosi/manifest.py [hunk 8]
```diff
@@ -197,6 +199,31 @@ def record_deb_packages(self, root: Path) -> None:
 
             source_package.add(package)
 
+    def record_pkg_packages(self, root: Path) -> None:
+        package_descriptions = [str(p) for p in (root / "/var/lib/pacman/local").glob("*/desc")]
```
[the remark this answers is not in the document]
Comment (behrmann) [reply mkosi/manifest.py]: Yes, with the `/` before `var` this ignores the initial `root`.

File: mkosi/manifest.py (hunk 7 above)
Comment (Foxboron) [reply mkosi/manifest.py:81]: Ahh, I need to get my Python-foo up to snuff. That is much better.

File: mkosi/manifest.py (hunk 6 above)
Comment (behrmann) [reply mkosi/manifest.py:219]: It takes some getting used to, but as soon as a Python codebase gets a sufficient size it *does* help to keep ones sanity. :)

File: mkosi/manifest.py (hunk 8 above)
[the remark this answers is not in the document]
Comment (Foxboron) [reply mkosi/manifest.py]: Yes, but note this comment is done on an outdated revision :)

I did notice this error myself while testing.

[the remark this answers is not in the document]
Comment (keszybz) [reply mkosi/manifest.py]: Yeah, I noticed I had some outdated comment when I was "publishing" the review. Github shows e.g. "2 comments", but if the comment is on a part of code that has been since changed, it doesn't show the comment anywhere. In the past it used to show all unpublished comments on the "comments" pane, but now it doesn't. So in the end I just pressed the green button without and forgot about this…

File: mkosi/manifest.py [hunk 9]
```diff
@@ -64,6 +64,20 @@ def report(self) -> str:
             t += f"""\nChangelog:\n{self.changelog}\n"""
         return t
 
+def parse_pkg_desc(f: Path) -> Tuple[str, str, str, str]:
+    with open(f) as desc:
```
Comment (keszybz) [inline mkosi/manifest.py:68]: I think `f.open()` is nicer, but OK.

**Comment by keszybz** [comment]
```
pyright 1.1.258
/home/runner/work/mkosi/mkosi/mkosi/manifest.py
  /home/runner/work/mkosi/mkosi/mkosi/manifest.py:80:12 - error: "name" is possibly unbound (reportUnboundVariable)
  /home/runner/work/mkosi/mkosi/mkosi/manifest.py:80:18 - error: "version" is possibly unbound (reportUnboundVariable)
  /home/runner/work/mkosi/mkosi/mkosi/manifest.py:80:27 - error: "base" is possibly unbound (reportUnboundVariable)
  /home/runner/work/mkosi/mkosi/mkosi/manifest.py:80:33 - error: "arch" is possibly unbound (reportUnboundVariable)
```

Meh, pyright strikes again. I guess this error needs to be silenced.

File: mkosi/manifest.py [hunk 10]
```diff
@@ -64,6 +64,20 @@ def report(self) -> str:
             t += f"""\nChangelog:\n{self.changelog}\n"""
         return t
 
+def parse_pkg_desc(f: Path) -> Tuple[str, str, str, str]:
+    with f.open() as desc:
```
Comment (behrmann) [inline mkosi/manifest.py:68]:
```suggestion

def parse_pkg_desc(f: Path) -> Tuple[str, str, str, str]:
    name = version = base = arch = ""
    with f.open() as desc:
```
Fixes the error pyright is complaining about and makes the code not break in case of a broken `desc` file.

File: mkosi/manifest.py [hunk 11]
```diff
@@ -64,6 +64,20 @@ def report(self) -> str:
...
+            if line == "%NAME%":
+                name = next(desc).strip()
+            elif line == "%VERSION%":
+                version = next(desc).strip()
+            elif line == "%BASE%":
+                base = next(desc).strip()
+            elif line == "%ARCH%":
+                arch = next(desc).strip()
+                break
+    return name, version, base, arch
```
Comment (behrmann) [inline mkosi/manifest.py:80]:
```suggestion
    return name, version, base, arch

```
while looking at what pyright is complaining about, I noticed missing newlines in between the functions. We follow a two newlines between functions style, like PEP8.

**Review by behrmann — APPROVED** [review]

**Review by DaanDeMeyer — APPROVED** [review]

**Comment by behrmann** [comment]
CI failure is unrelated. Merging.
